Show warning on AnimationTree for invalid advance expression

jackwilsdon opened the pull request:
<!--
Please target the `master` branch in priority.
PRs can target `3.x` if the same change was done in `master`, or is not relevant there.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.
You can mention in the description if the change is compatible with `3.x`.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->
This PR makes the engine display a warning in the scene tree editor against AnimationTrees which have a `AnimationNodeStateMachineTransition` with an invalid expression (i.e. one that can't be parsed):

![Configuration warning for invalid advance expression](https://github.com/godotengine/godot/assets/1843197/b30bd624-a746-46a9-96ff-f4b6bf6f198e)

It also adds some error reporting to `set_advance_expression`, which shows up in the editor's output (and debugger for running projects):

![Error in output](https://github.com/godotengine/godot/assets/1843197/d1ccc65a-9e9c-40ea-b06a-8ac46d96d085)

![Error in debugger](https://github.com/godotengine/godot/assets/1843197/b1e7a88f-c497-421a-bb6e-6546ec9f59a0)

One downside to adding this error reporting to the setter is that there is no debouncing when editing the expression via the inspector, causing every keystroke to report an error in the output (unless the expression is valid):

![Errors reported whilst typing "value == 3 + 4"](https://github.com/godotengine/godot/assets/1843197/8720fba6-576a-4c39-b7de-5609424b6b57)

I'm happy to remove this error reporting if need be, as it is only really useful when setting `advance_expression` at runtime, which I can't imagine is done very often.

This PR also adds error reporting to transition execution, catching cases where variables may not exist:
![Error at runtime](https://github.com/godotengine/godot/assets/1843197/4f97b533-4140-4fc8-9ed7-7822ed336d50)
There's a similar downside to this error reporting too - it ends up being output every frame that the expression is invalid (which is usually indefinitely), both in-editor and whilst the project is running. This could be changed to use `ERR_PRINT_ONCE`, but it seems like that would cause it to only be printed once per engine run, which seems less than ideal.

jackwilsdon committed:
Show warning on AnimationTree for invalid advance expression

jackwilsdon committed:
Don't report advance expression parse error in editor

We still report the error when running in a project in the editor, but
not when we're just in the editor. This avoids constantly printing
errors whilst the user is typing an expression.

jackwilsdon commented:
I've just pushed c396f5e2453b0e0aab83fcf0be749d3ffe720f73 which makes it so that we no longer report an error in the setter whilst running in the editor - this avoids the error being reported repeatedly whilst the user is typing in an expression. We still report an error if the project is running in the editor.

Calinou commented:
It may be worth detecting the `=` case in particular and suggesting the use of `==` in the node configuration warning.

jackwilsdon commented:
Yeah that sounds good. I've implemented it by checking the error returned by `get_error_text` - doing it this way seems a bit fragile, but Expression doesn't appear to expose anything else that we could check.

I've also changed the warning to now include the actual error returned by `get_error_text`, which should hopefully make it easier to understand:

![Configuration warning for invalid advance expression](https://github.com/godotengine/godot/assets/1843197/60b492db-4011-4371-8f96-0416c510f1b2)

The warning is getting a bit long now though - I'm wondering if it's worth dropping the actual expression from the error seeing as the warning already indicates which transition has the invalid expression.

Also, "maybe '=' should be '=='?" probably needs rewording, although I can't think of anything better off the top of my head.

jackwilsdon committed:
Add suggestion for "Expected '='" error

SaracenOne reviewed (changes requested):
I think this is fine since it provides much-needed feedback if something goes wrong, but I would recommend having the configuration_warnings bury into the subnodes too, since it currently only detects expression warnings in the highest level node, and its quite common to have a mix of blendtrees and state machines.
